url9ecf1e8297_nhamidi_ninja_tgz-pJ8-net_fec_openrq_test_TestRunnerJ8/src/RQLibrary/Encoder.java

```diff
@@ -0,0 +1,2552 @@
+	return pLinha; // T = P'
+    }
+    
+    private int derivateZ(int N_max) {
+	// Z = ceil(Kt/KL(N_max))
+	return (ceil((double) Kt / SystematicIndices.KL(N_max, MAX_SIZE_BLOCK, ALIGN_PARAM, T)));
+    }
+    
+    private int derivateN(int N_max) {
+	int n;
+	
+	// N is the minimum n=1, ..., N_max such that ceil(Kt/Z) <= KL(n)
+	for (n = 1; n <= N_max && ceil((double) Kt / Z) > SystematicIndices.KL(n, MAX_SIZE_BLOCK, ALIGN_PARAM, T); n++) {
+	}
+	
+	return n;
+    }
+    
+    public static final int ceil(double x) {
+	return ((int) Math.ceil(x));
+    }
+    
+    public static final int floor(double x) {
+	return ((int) Math.floor(x));
+    }
+    
+    public SourceBlock[] partition() {
+	/*
+	 * TODO is sub-blocking really necessary for regular use cases? (if it isn't, many mem copies could be avoided)
+	 * 
+	 * TODO give the user the option to use sub-blocking or not
+	 */
+	
+	// Partitioning parameters
+	
+	// (KL, KS, ZL, ZS) = Partition[Kt, Z]
+	Partition KZ = new Partition(Kt, Z);
+	int KL = KZ.get(1);
+	int KS = KZ.get(2);
+	int ZL = KZ.get(3);
+	
+	// (TL, TS, NL, NS) = Partition[T/Al, N]
+	Partition TN = new Partition(T / ALIGN_PARAM, N);
+	int TL = TN.get(1);
+	int TS = TN.get(2);
+	int NL = TN.get(3);
+	
+	// partitioned source blocks
+	SourceBlock[] object = new SourceBlock[Z];
+	
+	// source block index
+	int i;
+	
+	// original data buffer index
+	int index_master = 0;
+	
+	/*
+	 * the object MUST be partitioned into Z = ZL + ZS contiguous source blocks
+	 */
+	for (i = 0; i < ZL; i++) // first ZL
+	{ // Source block i
+	
+	    // allocate memory for source block i
+	    byte[] symbols = new byte[KL * T]; // a source block has KL*T octets
+	    
+	    // buffer index for source block i
+	    int index_symbols = 0;
+	    
+	    // we need this so we don't lose the point in index_master
+	    int aux_master = index_master;
+	    
+	    /*
+	     * each source block with K source symbols MUST be divided into N = NL + NS contiguous sub-blocks
+	     */
+	    for (int k = 0; k < KL; k++) {
+		// sub-block index
+		int j = 0;
+		
+		// where we will get the data from
+		int index_data = aux_master;
+		
+		/*
+		 * the first NL sub-blocks each consisting of K contiguous sub-symbols of size of TL*Al octets
+		 */
+		for (; j < NL; j++, index_data += KL * TL * ALIGN_PARAM, index_symbols += TL * ALIGN_PARAM) {
+		    System.arraycopy(data, index_data, symbols, index_symbols, TL * ALIGN_PARAM);
+		}
+		
+		/*
+		 * the remaining NS sub-blocks each consisting of K contiguous sub- symbols of size of TS*Al octets
+		 */
+		for (; j < N; j++, index_data += KL * TS * ALIGN_PARAM, index_symbols += TS * ALIGN_PARAM) {
+		    System.arraycopy(data, index_data, symbols, index_symbols, TS * ALIGN_PARAM);
+		}
+		
+		if ( NL > 0 ) {
+		    aux_master += TL * ALIGN_PARAM; // we always start at a sub-block of size TL*Al
+		} else {
+		    aux_master += TS * ALIGN_PARAM; // we always start at a sub-block of size TS*Al
+		}
+	    }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 184:210, 234:265.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.